Source/path.cpp

```diff
@@ -5,13 +5,426 @@
  */
 #include "path.h"
 
+#include <array>
+
 #include "gendung.h"
 #include "objects.h"
 
 namespace devilution {
 
 namespace {
 
+constexpr std::array<Displacement, 1369> CrawlOffsets {

```

Comment:
Extracting the counts was a nice improvement, but I would be nice if we could also make the bounds explicit instead of them just being arbitrary offsets in to a huge array.

Also what is left to be converted before we can fully drop the old array?

Reply:
The corners are the only exceptional case. If it wasn't slightly rounded we could just use the loop logic everywhere.

FireRing uses the rounding to make the effect look nice, all other uses of CrawlTable are used to find targets so squaring off the corners shouldn't cause visible impact. That would be StealPotions (used to select a target player), ChainLightning (used to select monsters to target, the visual effect is part of the missile movement logic) and the debug spawn monsters command (used to find free spaces).

Having a think about it, I can probably update the loop logic to inset the corners in the same way as the crawl table currently does so we could remove all the hardcoded values. This would mean that offsets +/-50, +/-50 are never checked but given searches that large are unlikely to have a completely occupied search space it shouldn't be a massive issue.